moveit_ros/moveit_servo/include/moveit_servo/servo_parameters.cpp

```diff
@@ -0,0 +1,286 @@
+/*********************************************************************
+ * Software License Agreement (BSD License)
+ *
+ *  Copyright (c) 2020, PickNik LLC
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions
+ *  are met:
+ *
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * Neither the name of PickNik LLC nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/* Author    : Adam Pettinger
+   Desc      : Declares, loads, and checks ServoParameters for Servo
+   Title     : servo_parameters.cpp
+   Project   : moveit_servo
+   Created   : 07/02/2020
+*/
+
+#include <rclcpp/rclcpp.hpp>
+
+#include <type_traits>
+
+#include <moveit_servo/servo_parameters.h>
+
+namespace moveit_servo
+{
+template <typename T>
+bool declareAndGetParam(T& output_value, const std::string& param_name, const rclcpp::Node::SharedPtr& node,
+                        const rclcpp::Logger& logger)
+{
+  bool got_param = false;
+  try
+  {
+    node->declare_parameter<T>(param_name, T{});
```

Comment:
I don't really understand why this should be necessary, but maybe I just didn't get the API right. I thought `declare_parameter()` already returns the correctly typed parameter value (default value or override from yaml) where trivial casts have already been applied (i.e. if you declare a double, an int should be casted to one).  If value conversion is not successful, `declare_parameter()` will throw an `InvalidParameterValueException`.  After declaring a parameter, `get_parameter()` will always return true (at least that's what I thought) so we would actually have to verify that the output value doesn't match the possibly invalid default value. The `false` case would only check for the `rclcpp::ParameterType::PARAMETER_NOT_SET` flag which is only used if undeclared parameters are allowed.

Reply:
I think you are right, I will look into this. In general I am unsure how to tell if a parameter was successfully set from yaml vs just taken as the default value. There are some parameters in `servo` that _**can**_ work as the default, but the user might not want that. And a bunch of course that cannot be default, and should be caught here as invalid